build.js

```diff
@@ -4,19 +4,17 @@ var fs = require('fs');
 var rimraf = require('rimraf');
 var acorn = require('acorn');
 var walk = require('acorn/dist/walk');
+var crypto = require('crypto');
 
-var ids = [];
-var names = {};
+var shasum = crypto.createHash('sha512');
+fs.readdirSync(__dirname + '/src').sort().forEach(function (filename) {
+  shasum.update(fs.readFileSync(__dirname + '/src/' + filename, 'utf8'));
+});
+var salt = shasum.digest('hex');
 
+// Deterministic 4 char suffix seeded from contents of src
 function getIdFor(name) {
-  if (name in names) return names[name];
-  var id;
-  do {
-    id = '_' + Math.floor(Math.random() * 100);
-  } while (ids.indexOf(id) !== -1)
-  ids.push(id);
-  names[name] = id;
-  return id;
+  return '_' + crypto.pbkdf2Sync(name, salt, 0, 2, 'sha512').toString('hex');
```

Comment:
Is it possible that two names could collide? 4 chars doesn't seem like very many but I also want to keep these identifiers as short as possible since minifiers rarely handle mangling property names.

We could keep the old code that caches the mapping from name to id, and keeps trying new ids until a unique one is found:

```js
  if (name in names) return names[name];
  var id;
  var keyLength = 1;
  do {
    id = '_' + crypto.pbkdf2Sync(
      name,
      salt,
      0, Math.ceil(keyLength / 2),
      'sha512'
    ).toString('hex').substr(0, keyLength);
    keyLength++;
  } while (ids.indexOf(id) !== -1)
  ids.push(id);
  names[name] = id;
  return id;
```

This would start us with one character identifiers, but simply add another character whenever a collision occurred, until a unique identifier was found.

Reply:
Mm, 16^4 == 65536 so 32768 with the birthday paradox. I think a cleaner solution than the re-use thing would be to increase the bytes but change the representation/base from `[0-9a-e]` to something like `[0-9a-zA-Z]`.

Reply:
I'd prefer to remove all risk of a collision rather than just adding entropy. Swapping to `[0-9a-zA-Z]` would be a good idea, but ideally I'd like it to use single character identifiers where possible. I remember it making a measurable difference to performance benchmarks when going from full names to these mangled, shorter names.

Reply:
What if we select one character in the range of `[a-z]` randomly to ensure no variable starts with a number, then append a base36 (`[0-9a-z]`) counter for each name we issue?

Reply:
@kpcyrd (That's not seeded on the contents of `src/`.)

Reply:
They need to start with `_` anyway to mark them as private.  It therefore doesn't need to start with a letter.  We could use:

```js
const names = {};
const characterSet = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
let i = characterSet.indexOf(shasum.digest('base64').replace(/[^0-9a-zA-Z]/g, '')[0]);

function getIdFor(name) {
  if (names[name]) return names[name];
  return names[name] = '_' + characterSet[i++ % characterSet.length]
}

This would be simpler than the what we have been using, and it would give us a single character identifier for up to 62 identifiers (and we definitely don't have anywhere near that many).